cub/cub/detail/nvtx.cuh

```diff
@@ -0,0 +1,114 @@
+/******************************************************************************
+ * Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+#pragma once
+
+#include <cub/config.cuh>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if __has_include(<source_location> )
+#  include <source_location>
+#endif // __has_include(<source_location>)
+
+// NVTX documentation: https://nvidia.github.io/NVTX/
+#include <cuda/std/optional>
+
+#include <nvtx3/nvToolsExt.h>
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+struct NVTXGlobal
+{
+  nvtxDomainHandle_t domain;
+
+  NVTXGlobal()
+  {
+    domain = nvtxDomainCreate("CUB");
+  }
+
+  ~NVTXGlobal()
+  {
+    nvtxDomainDestroy(domain);
+  }
```

Comment:
When I was writing the `nvtx3.hpp` C++ wrappers, I discussed with the NVTX folks that it was better for this destructor to do nothing. See https://github.com/NVIDIA/NVTX/blob/09e0d23a61ae86cc381eef85d012afc3a2e6eeea/c/include/nvtx3/nvtx3.hpp#L907-L921 

```suggestion
  ~NVTXGlobal() = default;
```

Reply:
Will do.

And about `nvtx3.hpp`: why is it not included in the CTK? I would like to use it for this PR. We discussed just pasting the header into the CCCL, but deemed it to be too large.

Reply:
It is included as of 12.3 iirc.

Reply:
I didn't find it in 12.4:
```
$ tree /usr/local/cuda-12.4/include/nvtx3
/usr/local/cuda-12.4/include/nvtx3
├── nvToolsExt.h
├── nvToolsExtCuda.h
├── nvToolsExtCudaRt.h
├── nvToolsExtOpenCL.h
├── nvToolsExtSync.h
└── nvtxDetail
    ├── nvtxImpl.h
    ├── nvtxImplCore.h
    ├── nvtxImplCudaRt_v3.h
    ├── nvtxImplCuda_v3.h
    ├── nvtxImplOpenCL_v3.h
    ├── nvtxImplSync_v3.h
    ├── nvtxInit.h
    ├── nvtxInitDecls.h
    ├── nvtxInitDefs.h
    ├── nvtxLinkOnce.h
    └── nvtxTypes.h

2 directories, 16 files
```